RFC: k8s / operator: offload CNPNodeStatus updates to cilium-operator

ianvernon opened the pull request:
**Problem Statement**
In large scale environments, the current model of having each node in the cluster try to update the NodeStatus for a CNP means that the amount of requests sent to update the status via the API server is a `multiple of number of in the cluster CNPs * number of nodes in the cluster` . This does not scale for large numbers of policies / large numbers of nodes in the cluster, as this results in a large amount of load on kube-apiserver. 

**Summary of Changes / Design**

To reduce the amount of interaction directly with the kube-apiserver, have each node update the key-value store used by Cilium with its status for the CNP and have `cilium-operator` be the sole interactor with kube-apiserver for updating the status. `cilium-operator` will watch for events from the key-value store for CNP statuses, and will then update the kube-apiserver with the updates.

The information is stored in the key-value store in the following format:

```
cilium/state/cnpstatuses/v2/<UUID of CNP structure>/<namespace>/<name>/<node name> --> <marshaled CNPNodeStatus + namespace, name, UID, node name> 
```

The operator has a kvstore watcher which spawns at bootstrap. Upon creation or modification of a key in the the `cilium/state/cnpstatuses/v2/` prefix, it will send the update to a goroutine which is responsible for updating the K8s apiserver with the status of a specific CNP.

The per-CNP updating management is performed by a goroutine which is launched when the operator receives a creation event for the CNP from its Kubernetes watcher, and stopped when the watcher receives a deletion event. On deletion, the prefix of `cilium/state/cnpstatuses/v2/<UUID of CNP structure>/<namespace>/<name>` is deleted from the key-value store. If a node is deleted, there is separate logic already present in `cilium-operator` which will GC old nodes from CNP NodeStatus.

Upon spawning of this goroutine, a shared-store which is launched at operator bootstrap is used to ensure that we do not miss any events which were pushed for CNPs which may have received the CNP creation event and pushed them to the key-value store before the operator received the CNP creation event from its own watcher.

These logic within the goroutine consumes all status events transmitted by the key-value store watcher until there are no more left available to consume. It pushes all of these updates into JSON Patch(es) and patches the CNP's status accordingly. This allows for multiple status updates to be sent to the kube-apiserver at once for multiple nodes as opposed to each node itself updating the kube-apiserver, improving scalability of NodeStatus updates. 

Fixes: #8897 

<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cilium/cilium/9384)
<!-- Reviewable:end -->

Commit:
k8s: refactor CNP update logic into separate function

No functional changes intended here. Future commits will rely upon this and
subsequent refactoring.

Signed-off by: Ian Vernon <ian@cilium.io>

Commit:
k8s: change `updateStatusByCapabilities` params

* Update to take node name, CNP name, client parameters
* Remove `CiliumUpdateStatusContext` parameter

Signed-off by: Ian Vernon <ian@cilium.io>

Commit:
k8s: return error if nil CNP is provided to `updateStatusWithCapabilities`

Signed-off by: Ian Vernon <ian@cilium.io>

Commit:
k8s: refactor `updateStatusByCapabilities`  to handle multiple node updates

Refactor the patching logic to be able to handle status updates for multiple
nodes as part of one update. No logical changes are intended here, as the only
path where this function is called right now only performs an update for one
node. Future commits will introduce callsites in `cilium-operator` which will
make use of the fact that multiple nodes can be updated for a single CNP.

Signed-off by: Ian Vernon <ian@cilium.io>

Commit:
k8s: factor out creation of CNPNodeStatus into function

There will be multiple creators of CNPNodeStatuses in different functions in
upcoming commits.

Signed-off by: Ian Vernon <ian@cilium.io>

aanm commented on pkg/k8s/cnp.go:
```diff
@@ -324,10 +324,10 @@ func (c *CNPStatusUpdateContext) update(cnp *types.SlimCNP, enforcing, ok bool,
 }
 
-func updateStatusByCapabilities(capabilities k8sversion.ServerCapabilities, c *CNPStatusUpdateContext, cnp *types.SlimCNP, ns string, cnpns cilium_v2.CiliumNetworkPolicyNodeStatus) error {
+func updateStatusByCapabilities(capabilities k8sversion.ServerCapabilities, c *CNPStatusUpdateContext, cnp *types.SlimCNP, ns, nodeName string, cnpns cilium_v2.CiliumNetworkPolicyNodeStatus) error {
```
If the node name is available in the CNPStatusUpdateContext why are we passing it as a parameter?

ianvernon replied:
See https://github.com/cilium/cilium/pull/9384#discussion_r333738610

aanm commented on pkg/k8s/cnp.go:
```diff
@@ -324,10 +324,10 @@ func (c *CNPStatusUpdateContext) update(cnp *types.SlimCNP, enforcing, ok bool,
 }
 
-func updateStatusByCapabilities(client clientset.Interface, capabilities k8sversion.ServerCapabilities, c *CNPStatusUpdateContext, cnp *types.SlimCNP, ns, nodeName string, cnpns cilium_v2.CiliumNetworkPolicyNodeStatus) error {
+func updateStatusByCapabilities(client clientset.Interface, capabilities k8sversion.ServerCapabilities, cnp *types.SlimCNP, ns, nodeName string, cnpns cilium_v2.CiliumNetworkPolicyNodeStatus) error {
```
This commit should be squash with the "k8s: change `updateStatusByCapabilities` to take node name parameter"

aanm commented on pkg/k8s/cnp.go:
```diff
@@ -324,10 +324,10 @@ func (c *CNPStatusUpdateContext) update(cnp *types.SlimCNP, enforcing, ok bool,
 }
 
-func updateStatusByCapabilities(client clientset.Interface, capabilities k8sversion.ServerCapabilities, cnp *types.SlimCNP, ns, nodeName string, cnpns cilium_v2.CiliumNetworkPolicyNodeStatus) error {
+func updateStatusByCapabilities(client clientset.Interface, capabilities k8sversion.ServerCapabilities, cnp *types.SlimCNP, ns, name, nodeName string, cnpns cilium_v2.CiliumNetworkPolicyNodeStatus) error {
```
If the name is available in the CNP why are we passing as a parameter?

ianvernon replied:
Future commits make this more clear - in the case of patching the CNP, we don't need to have the CNP itself at hand to perform the patch operation; we don't need to get it from the store from the SharedInformer. The status update value from the key-value store itself contains all the information we need about the CNP to update it. Do you recommend getting the value from the store unconditionally and using that instead even in the case where we have patching capabilities?

aanm commented on pkg/k8s/cnp.go:
```diff
@@ -383,15 +384,47 @@ func updateStatusByCapabilities(client clientset.Interface, capabilities k8svers
+		// Get the CNP from K8s so we can update its status. This case occurs
+		// when cilium-operator is updating the CNP and all it has is the
+		// CNP NodeStatus from the key-value store, but not the CNP itself.
+		if cnp == nil {
```
[1] This should be the cilium-operator's job. Not this function's job. Besides, cilium-operator should have a watcher listening for CNPs and it has access to the most recent revision of the CNP which does not require us to re-download the CNP again.

aanm commented on pkg/k8s/cnp.go:
```diff
@@ -383,15 +384,47 @@ func updateStatusByCapabilities(client clientset.Interface, capabilities k8svers
 		cnp.SetPolicyStatus(nodeName, cnpns)
 		_, err = client.CiliumV2().CiliumNetworkPolicies(ns).UpdateStatus(cnp.CiliumNetworkPolicy)
 	default:
+		// Get the CNP from K8s so we can update its status. This case occurs
```
same as [1]

ianvernon replied:
I changed the CNP watcher to use `NewInformerWithStore` like we have in `daemon/k8s_watcher.go`. I now get the CNP via `<namespace>/<name>` from this store if we do not have patching capabilities. LMK what you think.

aanm commented on pkg/k8s/cnp.go:
```diff
@@ -367,21 +367,58 @@ func updateStatusByCapabilities(client clientset.Interface, capabilities k8svers
-			createStatusAndNodePatchJSON, err = json.Marshal(createStatusAndNodePatch)
-			if err != nil {
-				return err
+			// If there are more than MaxJSONPatchOperations to patch, do
```
Since we have a single entity updating the CNP status, the cilium-operator, why don't we patch the full `/status/nodes` instead of patching each individual one?

ianvernon replied:
My motivation for doing it this way was to only have us replace the node status updates for the nodes which we have received events for. We would have to maintain the state of all NodeStatuses for the given CNP (which we already have in the map of node name --> NodeStatus passed into this function) and put that entire map into the PATCH every time we update it, right? My intuition was that this would lead to us sending unnecessary information to the K8s apiserver if say there was only one node that was updated, as we would only transmit the patch for that node, not all prior nodes + this new node. But, it would reduce the total number of patches that we would be transmitting. I don't have any hard data to back this up, however. I can try measuring both? What do you think? I am open to all suggestions here, and you have a lot of prior experience with optimizing K8s apiserver interactions.

ianvernon replied:
Newest set of commits is using MergePatch instead of just Patch, still WIP though. Need to run in case where offload is enabled vs. not enabled manually on my local setup.

ianvernon replied:
Unfortunately, using merge+patch will not work with our status structure, because of the fact that the `Annotations` field within the NodeStatus has a `map` within it. The way MergePatch works with maps, is that it only overwrites the key-value pairs present in the MergePatch; this works will with the mapping of nodes --> NodeStatus, but within the NodeStatus structure, does not work well for annotations. If we apply an initial MergePatch with the following annotations:

```
foo --> bar
```

and then apply another update which we want to *replace* the annotations with 

```
baz --> bum
```

the annotations in K8s will appear as:

```
foo --> bar
baz --> bum
```

aanm commented on pkg/k8s/cnp.go:
```diff
@@ -269,12 +272,52 @@ retryLoop:
-func (c *CNPStatusUpdateContext) update(cnp *types.SlimCNP, enforcing, ok bool, cnpError error, rev uint64, cnpAnnotations map[string]string) error {
+// CNPStatusesPath is the prefix in the kvstore which will contain all keys
+// representing CNPStatus state for all nodes in the cluster.
+var CNPStatusesPath = path.Join(kvstore.BaseKeyPrefix, "state", "cnpstatuses", "v1")
```
the k8s CNP is a v2 object

ianvernon replied:
Fixed in follow-up commit.

aanm commented on pkg/k8s/cnp.go:
```diff
@@ -307,9 +350,26 @@ func (c *CNPStatusUpdateContext) update(cnp *types.SlimCNP, enforcing, ok bool,
+			"key":   key,
+			"value": marshaledVal,
+		}).Debug("updating CNPStatus in kvstore")
+		_, err = kvstore.Client().UpdateIfDifferent(m.ctx, key, marshaledVal, true)
```
In this case I don't think it matter, IIRC `UpdateIfDifferent` is really useful to avoid triggering KVStore events to all watchers. In this case there's only one watcher which is the operator.

ianvernon replied:
Fixed in follow-up commit.

aanm commented on operator/cnp_event.go:
```diff
@@ -16,15 +16,22 @@ package main
 
 import (
 	"context"
+	"encoding/json"
```
I only read the commit description here, probably it makes sense to have the 10s as a flag for cilium operator? :thinking:

ianvernon commented on operator/cnp_event.go:
```diff
@@ -101,3 +130,179 @@ func enableCNPWatcher() error {
+						StopFunc: func(ctx context.Context) error {
+							close(ch)
+							mgr.mutex.Lock()
+							delete(mgr.eventMap, nameNamespace)
```
I have a feeling that this isn't as simple as removing from the map, as I'm not sure what's stopping a `Create` or `Modify` event from coming in from another node for this same CNP and adding it back into the controller manager after we have stopped this controller in the CNP watcher above...

I was thinking maybe we should wait for a delete event to be received from every single node until there are no more keys left for this CNP in the key-value store (tracked by perhaps a set of all node names that this CNP is aware of), and compare that to the set of nodes in the cluster at the time we receive the delete event from K8s?

ianvernon commented on operator/cnp_event.go:
```diff
@@ -101,3 +130,179 @@ func enableCNPWatcher() error {
+				nameNamespace := fmt.Sprintf("%s/%s", namespace, name)
+				mgr.mutex.Lock()
+				ch, ok := mgr.eventMap[nameNamespace]
+				if !ok {
```
@aanm would checking if the CNP for which I received the update is in the `cache.Store` resolve the potential race [here](https://github.com/cilium/cilium/pull/9384/files#r333750774) ? Let me know if you need me to elaborate / explain what I am thinking of - I had a bit of trouble putting it into words when thinking of this.

ianvernon commented on operator/main.go:
```diff
@@ -312,6 +312,7 @@ func runOperator(cmd *cobra.Command) {
 	if identityGCInterval != time.Duration(0) {
 		startIdentityGC()
 	}
+
```
Remove this.

Commit:
k8s: send CNP status to kvstore if handover enabled

If Kubernetes event handover to the key-value store is enabled, then have each
cilium-agent send the status update to the key-value store. The cilium-operator
will set up a watcher which will watch for these events, and will be the sole
entity interacting with the Kubernetes apiserver with respect to CNP NodeStatus
updates.

Signed-off by: Ian Vernon <ian@cilium.io>

Commit:
operator: send CNPNodeStatus updates to k8s

In large scale environments, the current model of having each node in the
cluster try to update the NodeStatus for a CNP means that the amount of requests
sent to update the status via the API server is the following product:

```
\# of CNPs * # of nodes in the cluster
```

This does not scale for large numbers of policies / large numbers of nodes in
the cluster, as this results in a large amount of load on kube-apiserver.

To reduce the amount of interaction directly with the kube-apiserver, have each
ode update the key-value store used by Cilium with its status for the CNP and
have `cilium-operator` be the sole interactor with kube-apiserver for updating
the status. `cilium-operator` will watch for events from the key-value store for
CNP statuses, and will then update the kube-apiserver with the updates.

The information is stored in the key-value store in the following format:

```
cilium/state/cnpstatuses/v2/<UUID of CNP structure>/<namespace>/<name>/<node name> --> <marshaled CNPNodeStatus + namespace, name, UID, node name>
```

The operator has a kvstore watcher which spawns at bootstrap. Upon creation or
modification of a key in the the `cilium/state/cnpstatuses/v2/` prefix, it will
send the update to a goroutine which is responsible for updating the K8s
apiserver with the status of a specific CNP.

The per-CNP updating management is performed by a goroutine which is launched
when the operator receives a creation event for the CNP from its Kubernetes
watcher, and stopped when the watcher receives a deletion event. On deletion,
the prefix of
`cilium/state/cnpstatuses/v2/<UUID of CNP structure>/<namespace>/<name>` is
deleted from the key-value store. If a node is deleted, there is separate logic
already present in `cilium-operator` which will GC old nodes from CNPNodeStatus.

Upon spawning of this goroutine, a shared-store which is launched at operator
bootstrap is used to ensure that we do not miss any events which were pushed for
CNPs which may have received the CNP creation event and pushed them to the
key-value store before the operator received the CNP creation event from its own
watcher.

The logic within the goroutine per-CNP consumes all status events transmitted by
the key-value store watcher until a configurable timeout
(`cnp-status-update-interval`, default 1s) is hit. It pushes all of these
updates to kube-apiserver in the most efficient / performant way depending on
which Kubernetes version is running. For >=1.13, JSON Patch(es) are used.

Using the cilium-operator to perform these updates allows for multiple status
updates to be sent to the kube-apiserver at once for multiple nodes as opposed
to each node itself updating the kube-apiserver, improving scalability of
CNPNodeStatus updates.

Signed-off by: Ian Vernon <ian@cilium.io>

aanm commented on pkg/k8s/cnp.go:
```diff
@@ -304,27 +304,30 @@ func (c *CNPStatusUpdateContext) update(cnp *types.SlimCNP, enforcing, ok bool,
+		OK:          ok,
+		LastUpdated: cilium_v2.NewTimestamp(),
+		Annotations: annotations,
+	}
```
I would suggest this to live as a `CiliumNetworkPolicyNodeStatus` constructor in `cilium_v2` package.

Also, maybe to decrease the LOC
```golang
func createCNPNodeStatus(enforcing, ok bool, cnpError error, rev uint64, annotations map[string]string) cilium_v2.CiliumNetworkPolicyNodeStatus {
	cnpns := cilium_v2.CiliumNetworkPolicyNodeStatus{
		Enforcing:   enforcing,
		OK:          ok,
		LastUpdated: cilium_v2.NewTimestamp(),
		Annotations: annotations,
	}
	if cnpError != nil {
		cnpns.Error = cnpError.Error()
		return cnpns
	}
	cnpns.Revision = rev
	return cnpns
}
```

aanm commented on pkg/k8s/cnp.go:
```diff
@@ -268,12 +285,52 @@ retryLoop:
+
+// UpdateMetadata contains information about how the update for the NodeStatus
+// in a CNP will be propagated.
+type UpdateMetadata struct {
```
What are we gaining for having this structure? This seems to be hiding the "context" away of the function arguments. I would suggest to have 2 functions that perform the update, one called "updateToKVStore" and other called "updateToAPIServer". The code duplication of having 2 functions is not big enough that would justify having 2 switch cases in `update`

ianvernon replied:
Posted a commit which removes this.

aanm commented on pkg/k8s/cnp.go:
```diff
@@ -306,9 +363,67 @@ func (c *CNPStatusUpdateContext) update(cnp *types.SlimCNP, enforcing, ok bool,
+		ns := k8sUtils.ExtractNamespace(&cnp.ObjectMeta)
+		return updateStatusesByCapabilities(c.CiliumNPClient, capabilities, cnp, ns, cnp.GetName(), map[string]cilium_v2.CiliumNetworkPolicyNodeStatus{c.NodeName: cnpns})
+	case KVStore:
+		cnpWithMeta := CNPNSWithMeta{
```
`CiliumNetworkPolicyNodeStatus` is the only value we need. All of the other fields (name, namespace, uid, nodename) are part of the key name as formatted in line 383

ianvernon replied:
I found it cleaner to be able to unmarshal and access the keys instead of having to do string manipulation - is the motivation behind this to avoid duplication of data stored in the key value store?

aanm replied:
yes

ianvernon replied:
Actually, I remember why I had to do it this way - I need to be able to implement `GetKeyName` to implement the `Key` interface so that I can store key-value pairs in the `SharedStore`. So, I need to encapsulate all of the information which is needed for the key corresponding to the value, within the value itself - the UUID/Namespace/Name/NodeName.

aanm commented on pkg/k8s/cnp.go:
```diff
@@ -306,9 +363,67 @@ func (c *CNPStatusUpdateContext) update(cnp *types.SlimCNP, enforcing, ok bool,
+			CiliumNetworkPolicyNodeStatus: cnpns,
+			Node:                          node.GetName(),
+		}
+		marshaledVal, err := json.Marshal(cnpWithMeta)
```
```patch
@@ -368,14 +368,7 @@ func (c *CNPStatusUpdateContext) update(m *UpdateMetadata, cnp *types.SlimCNP, e
                ns := k8sUtils.ExtractNamespace(&cnp.ObjectMeta)
                return updateStatusesByCapabilities(c.CiliumNPClient, capabilities, cnp, ns, cnp.GetName(), map[string]cilium_v2.CiliumNetworkPolicyNodeStatus{c.NodeName: cnpns})
        case KVStore:
-               cnpWithMeta := CNPNSWithMeta{
-                       Name:                          cnp.GetName(),
-                       Namespace:                     cnp.GetNamespace(),
-                       UID:                           cnp.GetUID(),
-                       CiliumNetworkPolicyNodeStatus: cnpns,
-                       Node:                          node.GetName(),
-               }
-               marshaledVal, err := json.Marshal(cnpWithMeta)
+               marshaledVal, err := json.Marshal(cnpns)
                if err != nil {
                        return err
                }
```

aanm commented on pkg/k8s/cnp.go:
```diff
@@ -448,6 +565,8 @@ func updateStatusesByCapabilities(client clientset.Interface, capabilities k8sve
 	// Updating succeeded - the updated map can be 'emptied' of updates that
 	// we need to propagate.
-	nodeStatuses = map[string]cilium_v2.CiliumNetworkPolicyNodeStatus{}
+	for k := range nodeStatuses {
+		delete(nodeStatuses, k)
+	}
```
Why this change?

ianvernon replied:
`ineffassign` complains that I am doing an ineffectual assignment (which surprised me, since I am updating the map itself which is a reference). The compiler will optimize the "for all entries, delete" FWIW.

aanm commented on pkg/k8s/cnpstatus.go:
```diff
@@ -0,0 +1,301 @@
+				// TODO - the channel may block once full, which means that
+				// we would potentially block for up to 10 seconds (run of the
+				// controller) before consuming the next event.
+				ch <- nsu
```
If `!ok` this line will block forever

aanm replied:
[1] panic

aanm commented on pkg/k8s/cnpstatus.go:
```diff
@@ -0,0 +1,301 @@
+							log.Errorf("received unexpected type mapping to key %s in cnp shared store: %T", k, sharedKeys[k])
+							continue
+						}
+						nodeStatusMap[v.Node] = v.CiliumNetworkPolicyNodeStatus
```
```patch
-                               for k := range sharedKeys {
+                               for keyName, storeKey := range sharedKeys {
                                        // Look for any key which matches this CNP.
-                                       if strings.HasPrefix(k, cnpKey) {
-                                               v, ok := sharedKeys[k].(*CNPNSWithMeta)
+                                       if strings.HasPrefix(keyName, cnpKey) {
+                                               cnpns, ok := storeKey.(*cilium_v2.CiliumNetworkPolicyNodeStatus)
                                                if !ok {
-                                                       log.Errorf("received unexpected type mapping to key %s in cnp shared store: %T", k, sharedKeys[k])
+                                                       log.Errorf("received unexpected type mapping to key %s in cnp shared store: %T", keyName, storeKey)
                                                        continue
                                                }
-                                               nodeStatusMap[v.Node] = v.CiliumNetworkPolicyNodeStatus
+                                               // extract nodeName from keyName
+                                               nodeStatusMap[nodeName] = cnpns
```

aanm commented on pkg/k8s/cnpstatus.go:
```diff
@@ -0,0 +1,301 @@
+				case <-ctx.Done():
+					// Controller was stopped, we can simply exit.
+					return nil
+				case ev, ok := <-ch:
```
Why are we still doing any kind of work if `!ok`?

aanm commented on pkg/k8s/cnpstatus.go:
```diff
@@ -0,0 +1,301 @@
+			} else {
+				// Wait for one status update to occur.
+				select {
+				case ev, ok := <-ch:
```
Why are we still doing any kind of work if `!ok`?

aanm commented on pkg/k8s/cnpstatus.go:
```diff
@@ -0,0 +1,301 @@
+			// at least one event in the first run.
+			if len(nodeStatusMap) == 0 {
+				return nil
+			}
```
I don't understand this. This seems to be solving something that can be avoided with a rewrite of this function's logic.

ianvernon replied:
My goal here was to ensure that we get all updates that came before we started this goroutine from the SharedStore, but that if we have received updates from the shared store, to not block on receiving at least one node status update from the channel, because we already "received" them from the store; during the first run, there are two sources to 'receive' updates - the map is the 'receiver' of these updates, so checking whether the map has been updated seems like a reasonable check to me here. If more updates are sent after we start the goroutine (by reading from the channel on line 235), we will batch those and send them to the K8s apiserver all in one go. In all other cases, except for the first run, we wait for at least one event to be received, and then wait for subsequent events or until no more subsequent events are received at some time.

ianvernon replied:
Next push will have this explained more clearly / performed more cleanly.

aanm commented on pkg/k8s/cnpstatus.go:
```diff
@@ -0,0 +1,301 @@
+		},
+		RunInterval: c.updateInterval,
+		StopFunc: func(ctx context.Context) error {
+			close(ch)
```
Closing this here will make panics in [1]

ianvernon replied:
I've added a channel which is closed upon receiving a delete event for a given CNP. The kvstore watcher, which is the sender to the status update channel, now also closes the channel if this shutdown channel is closed, ensuring no panics will occur. The goroutine which is responsible for receiving status update events also will exit if this channel is closed.

aanm commented on pkg/k8s/cnp.go:
```diff
@@ -361,33 +485,88 @@ func (c *CNPStatusUpdateContext) update(cnp *types.SlimCNP, enforcing, ok bool,
+				// to update.
+				for i := range nodeNamesUsed {
+					delete(nodeStatuses, nodeNamesUsed[i])
+				}
```
```patch
				for _, nodeName := range nodeNamesUsed {
					delete(nodeStatuses, nodeName)
				}
```

aanm commented on pkg/k8s/cnp.go:
```diff
@@ -293,42 +288,58 @@ func formatKeyForKvstore(uid k8sTypes.UID, namespace, name, nodeName string) str
+		annotations = cnpAnnotations
+		lastAppliedConfig, ok := annotations[v1.LastAppliedConfigAnnotation]
+		defer func() {
+			if ok {
```
run the defer if ok, not the other way around.

aanm commented on pkg/k8s/cnpstatus.go:
```diff
@@ -0,0 +1,345 @@
+				}
+				nodeStatusMap[ev.node] = *ev.CiliumNetworkPolicyNodeStatus
+			}
+		}
```
```golang
	for {
		limit := time.After(c.updateInterval)

		// Collect any other events that have come in, but bail out after the
		// above limit is hit so that we can send the updates we have received.
	Loop:
		for {
			select {
			case <-nodeStatusUpdater.stopChan:
				return
			case <-limit:
				if len(nodeStatusMap) == 0 {
					// If nothing to update, wait until we have something to update.
					limit = nil
					continue
				}
				break Loop
			case ev, ok := <-nodeStatusUpdater.updateChan:
				if !ok {
					return
				}
				nodeStatusMap[ev.node] = *ev.CiliumNetworkPolicyNodeStatus
			}
		}
```

aanm commented on operator/cnp_event.go:
```diff
@@ -41,7 +49,38 @@ func init() {
+		Prefix: k8s.CNPStatusesPath,
+		KeyCreator: func() store.Key {
+			newCNPNS := k8s.CNPNSWithMeta{}
+			return &newCNPNS
```
```golang
return &k8s.CNPNSWithMeta{}
```

aanm commented on pkg/k8s/cnpstatus.go:
```diff
@@ -0,0 +1,345 @@
+
+				// Send the update to the corresponding controller for the
+				// CNP which sends all status updates to the K8s apiserver.
+				cnpKey := fmt.Sprintf("%s/%s/%s", cnpStatusUpdate.UID, cnpStatusUpdate.Namespace, cnpStatusUpdate.Name)
```
I think it's better to create a method that generates the key. I've seen similar code in different places which can be error prone. Besides, this key is different than the one in the commit message:

```
cilium/state/cnpstatuses/v2/<UUID of CNP structure>/<namespace>/<name>/<node name> -
```

ianvernon replied:
Agreed re: having a function. However, the commit message refers to the format in the key-value store. The format in the key you refer to is not for the key-value store.

aanm commented on operator/main.go:
```diff
@@ -150,6 +150,8 @@ func init() {
 	flags.BoolVar(&enableCNPNodeStatusGC, "cnp-node-status-gc", true, "Enable CiliumNetworkPolicy Status garbage collection for nodes which have been removed from the cluster")
 	flags.DurationVar(&ciliumCNPNodeStatusGCInterval, "cnp-node-status-gc-interval", time.Minute*2, "GC interval for nodes which have been removed from the cluster in CiliumNetworkPolicy Status")
 
+	flags.DurationVar(&cnpStatusUpdateInterval, "cnp-status-update-interval", 100*time.Millisecond, "interval between CNP status updates sent to the k8s-apiserver per-CNP")
```
Is this time correct? 100 milliseconds?

ianvernon replied:
Yes. This is used to ratelimit the number of updates to K8s apiserver per CNP to at minimum 100 milliseconds. I figured it was a sane default. Is there a specific problem you see with this value?

aanm replied:
This means we will push every 100 milliseconds, I think it's too aggressive. We would need to get more data to get a sane value, but given the status update can will be read by humans I think they won't be checking every 100 miliseconds to see if the status has changed. Perhaps 1 seconds should be sufficient.

ianvernon replied:
Changed to 1 second.

aanm commented on pkg/k8s/cnpstatus.go:
```diff
@@ -0,0 +1,335 @@
+	// Signal that we should stop processing events.
+	close(nsu.stopChan)
+	delete(c.eventMap, cnpKey)
+	c.Unlock()
```
use defer to avoid the bug in line 96

ianvernon replied:
Fixed.

aanm commented on pkg/k8s/cnpstatus.go:
```diff
@@ -0,0 +1,335 @@
+		stopChan:   make(chan struct{}),
+	}
+	c.eventMap[cnpKey] = nsu
+	c.Unlock()
```
use defer

ianvernon replied:
Fixed.

aanm reviewed (approved):
LGTM. I'm not sure if you want to squash any commits before merging so I'm leaving the PR open for now.

joestringer reviewed (approved):
Docs LGTM.
